packages/features/schedules/lib/use-schedule/useNonEmptyScheduleDays.ts

```diff
@@ -1,23 +1,79 @@
 import { useMemo } from "react";
 
+import type { ManipulateType } from "@calcom/dayjs";
+import dayjs from "@calcom/dayjs";
+import { yyyymmdd } from "@calcom/lib/date-fns";
+import { Frequency } from "@calcom/prisma/zod-utils";
+
 import type { Slots } from "../use-schedule";
 
-export const getNonEmptyScheduleDays = (slots?: Slots) => {
-  if (typeof slots === "undefined") return [];
+type AdditionalParams = {
+  occurenceCount?: null | number;
+  recurringEventFreq?: Frequency;
+};
+
+const retainDaysWithFirstTwoSlotsAvailable = (
+  nonEmptyDays: Set<string>,
+  occurenceCount: number,
+  recurringEventFreq: Frequency
+) => {
+  const isDateSlotAvailable = (date: string) => nonEmptyDays.has(date);
+
+  // If occurenceCount is 1, no modifications are needed
+  if (occurenceCount === 1) return;
+
+  const freqKey = Frequency[recurringEventFreq]?.toLowerCase();
+
+  // Only process supported frequency types
+  if (!["weekly", "monthly", "yearly"].includes(freqKey)) return;
 
-  const nonEmptyDays: string[] = [];
+  const frequencyToSuffix: Record<string, ManipulateType> = {
+    weekly: "week",
+    monthly: "month",
+    yearly: "year",
+  };
 
+  const suffix = frequencyToSuffix[freqKey];
+
+  if (suffix) {
+    nonEmptyDays.forEach((date) => {
+      const nextDateSlot = yyyymmdd(dayjs(date).add(1, suffix));
+      if (!isDateSlotAvailable(nextDateSlot)) {
+        nonEmptyDays.delete(date);
+      }
+    });
+  }
+};
+
+export const getNonEmptyScheduleDays = (slots?: Slots, additionalParams?: AdditionalParams) => {
+  if (!slots || typeof slots === "undefined") return [];
+  const nonEmptyDays = new Set<string>();
+
+  // Populate initial set of non-empty days
   Object.keys(slots).forEach((date) => {
     if (slots[date].some((slot) => !(slot?.away && !slot.toUser) && slots[date].length > 0)) {
-      nonEmptyDays.push(date);
+      nonEmptyDays.add(date);
     }
   });
 
-  return nonEmptyDays;
-};
+  // For recurring events
+  if (
+    additionalParams?.occurenceCount &&
+    (additionalParams?.recurringEventFreq || additionalParams?.recurringEventFreq == 0)
```

Comment:
This check seems tricky, would it make sense to move it to `retainDaysWithFirstTwoSlotsAvailable` fn itself? 

We could just keep `additionalParams?.occurenceCount` here and rest inside the fn